File: ts/project/.sdk/tm/lua/test/omni_smoke_test.lua
-- ProjectName SDK omni runner smoke test
--
-- Smoke tests for the vendored omni runner itself: a runner that cannot
-- FAIL a bad entry would turn every corpus suite vacuously green, so pin
-- the failure paths, not just the happy one. (Lua peer of ts's
-- test/omni.test.ts and py's test_omni_smoke.py.)

local assert = require("luassert")

local sdk = require("project-name-_sdk")
local omni = require("test.omni")

local makeRunner, isomnierror = omni.makeRunner, omni.isomnierror
local map, list = omni.map, omni.list


-- A minimal in-memory spec: no fixture file, no OMNI block (lenient v0,
-- like the shared corpus). Built in omni's own value model - a plain lua
-- table is invisible to its walkers.
local function makespec()
  return map({
    primary = map({
      smoke = map({
        basic = map({
          set = list({
            map({ ["in"] = 1, out = 2 }),
            map({ ["in"] = 41, out = 42 }),
          }),
        }),
        bad = map({
          set = list({
            map({ ["in"] = 1, out = 999 }),
          }),
        }),
        err = map({
          set = list({
            map({ ["in"] = 0, err = "zero refused" }),
          }),
        }),
      }),
    }),
  })
end


local function inc(n)
  if 0 == n then
    error("smoke: zero refused")
  end
  return n + 1
end


-- Run `fn`, requiring it to raise an omni failure whose message contains
-- `want`.
local function mustfail(fn, want)
  local ok, err = pcall(fn)
  assert.is_false(ok, "expected the runner to FAIL, but it passed")
  assert.is_true(isomnierror(err),
    "expected an omni error, got: " .. tostring(err))
  assert.is_truthy(string.find(tostring(err), want, 1, true),
    "expected failure containing '" .. want .. "', got: " .. tostring(err))
end


describe("omni smoke", function()

  local function pack()
    local runner = makeRunner(makespec(), sdk.test(nil, nil))
    return runner("smoke")
  end


  it("runset passes a correct subject", function()
    local R = pack()
    R.runset(R.spec.basic, inc)
  end)


  it("runset fails a wrong result with an omni error", function()
    local R = pack()
    mustfail(function()
      R.runset(R.spec.bad, inc)
    end, "result mismatch")
  end)


  it("an expected error is matched, and a missing expected error fails", function()
    local R = pack()

    -- The expected error occurs: passes.
    R.runset(R.spec.err, inc)

    -- The expected error does NOT occur: must fail.
    local R2 = pack()
    mustfail(function()
      R2.runset(R2.spec.err, function(n)
        return n
      end)
    end, "expected error did not occur")
  end)

end)

File: ts/project/.sdk/tm/elixir/test/omni_smoke_test.exs
# ProjectName SDK omni runner smoke test
#
# Smoke tests for the VENDORED omni runner itself. A runner that cannot FAIL
# a bad entry would turn every corpus suite vacuously green, so the failure
# paths are pinned here, not just the happy one. (Elixir peer of ts's
# test/omni.test.ts, lua's test/omni_smoke_test.lua and py's
# test_omni_smoke.py.)
#
# The spec is built IN MEMORY, in omni's own value model (plain maps and
# lists, no OMNI block - the lenient v0 format the shared corpus uses), so
# these tests depend on no fixture file.

defmodule ProjectName.OmniSmokeTest do
  use ExUnit.Case

  alias ProjectName.Omni, as: O

  @spec_smoke %{
    "primary" => %{
      "smoke" => %{
        "basic" => %{
          "set" => [
            %{"in" => 1, "out" => 2},
            %{"in" => 41, "out" => 42}
          ]
        },
        "bad" => %{
          "set" => [
            %{"in" => 1, "out" => 999}
          ]
        },
        "err" => %{
          "set" => [
            %{"in" => 0, "err" => "zero refused"}
          ]
        },
        "argsback" => %{
          "set" => [
            %{"in" => %{"n" => 1}, "match" => %{"args" => %{"0" => %{"n" => 2}}}}
          ]
        }
      }
    }
  }

  # The subject contract the resolver presents: omni's resolved argument list,
  # in the SDK's value model.
  defp inc(args) do
    n = hd(args)
    if n == 0, do: raise(Voxgig.Struct.Error, message: "smoke: zero refused")
    n + 1
  end

  defp pack do
    runner = O.make_runner(@spec_smoke, ProjectName.test())
    runner.("smoke")
  end

  defp mustfail(fun, want) do
    err =
      try do
        fun.()
        flunk("expected the runner to FAIL, but it passed")
      rescue
        e -> e
      end

    assert O.omnierror?(err), "expected an omni error, got: #{inspect(err)}"

    assert String.contains?(Exception.message(err), want),
           "expected a failure containing '#{want}', got: #{Exception.message(err)}"
  end

  test "runset passes a correct subject" do
    run = pack()
    run.runset.(run.spec["basic"], &inc/1)
    assert O.cases() > 0
  end

  test "runset fails a wrong result with an omni error" do
    run = pack()
    mustfail(fn -> run.runset.(run.spec["bad"], &inc/1) end, "result mismatch")
  end

  test "an expected error is matched, and a missing expected error fails" do
    run = pack()

    # The expected error occurs: passes.
    run.runset.(run.spec["err"], &inc/1)

    # The expected error does NOT occur: must fail.
    run2 = pack()
    mustfail(fn -> run2.runset.(run2.spec["err"], fn args -> hd(args) end) end,
      "expected error did not occur")
  end

  test "an in-place argument rewrite reaches match.args" do
    run = pack()

    # The writeback channel (omni.ex decision 2): the subject MUTATES its
    # argument node, and `match: {args: ...}` must see it. A resolver that
    # dropped the channel would read the pre-call value and this passes only
    # by accident, so the negative case is pinned too.
    run.runset.(run.spec["argsback"], fn args ->
      node = hd(args)
      Voxgig.Struct.setprop(node, "n", 2)
      nil
    end)

    run2 = pack()

    mustfail(
      fn -> run2.runset.(run2.spec["argsback"], fn _args -> nil end) end,
      "match failed at args.0.n"
    )
  end
end
